src/user-pool.ts

```diff
@@ -0,0 +1,572 @@
+   * @default - no trigger configured
+   */
+  readonly userMigration?: Trigger<lambda.UserMigrationTriggerEvent>;
+  /**
+   * Verifies the authentication challenge response.
+   *
+   * ```ts
+   * new UserPool(scope, id, {
+   *   verifyAuthChallengeResponse: new Function(scope, id, async (event) => {
+   *     // handle event, set response properties, etc.
+   *     event.response... = ..
+   *     return event;
+   *   })
+   * });
+   * ```
+   *
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-verify-auth-challenge-response.html
+   * @default - no trigger configured
+   */
+  readonly verifyAuthChallengeResponse?: Trigger<lambda.VerifyAuthChallengeResponseTriggerEvent>;
+  /**
+   * Amazon Cognito invokes this trigger to send email notifications to users.
+   *
+   * ```ts
+   * new UserPool(scope, id, {
+   *   customEmailSender: new Function(scope, id, async (event) => {
+   *     // handle event, set response properties, etc.
+   *     event.response... = ..
+   *     return event;
+   *   })
+   * });
+   * ```
+   *
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-email-sender.html
+   * @default - no trigger configured
+   */
+  readonly customEmailSender?: Trigger<lambda.CustomEmailSenderTriggerEvent>;
+  /**
+   * Amazon Cognito invokes this trigger to send SMS notifications to users.
+   *
+   * ```ts
+   * new UserPool(scope, id, {
+   *   customSmsSender: new Function(scope, id, async (event) => {
+   *     // handle event, set response properties, etc.
+   *     event.response... = ..
+   *     return event;
+   *   })
+   * });
+   * ```
+   *
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-sms-sender.html
+   * @default - no trigger configured
+   */
+  readonly customSmsSender?: Trigger<lambda.CustomSMSSenderTriggerEvent>;
+}
+
+/**
+ * Map a friendly trigger name to the {@link aws_cognito.UserPoolOperation}.
+ */
+const triggers: {
+  [triggerName in keyof UserPoolTriggers]: aws_cognito.UserPoolOperation;
+} = {
+  createAuthChallenge: aws_cognito.UserPoolOperation.CREATE_AUTH_CHALLENGE,
+  customEmailSender: aws_cognito.UserPoolOperation.CUSTOM_EMAIL_SENDER,
+  customMessage: aws_cognito.UserPoolOperation.CUSTOM_MESSAGE,
+  customSmsSender: aws_cognito.UserPoolOperation.CUSTOM_SMS_SENDER,
+  defineAuthChallenge: aws_cognito.UserPoolOperation.DEFINE_AUTH_CHALLENGE,
+  postAuthentication: aws_cognito.UserPoolOperation.POST_AUTHENTICATION,
+  postConfirmation: aws_cognito.UserPoolOperation.POST_CONFIRMATION,
+  preAuthentication: aws_cognito.UserPoolOperation.PRE_AUTHENTICATION,
+  preSignUp: aws_cognito.UserPoolOperation.PRE_SIGN_UP,
+  preTokenGeneration: aws_cognito.UserPoolOperation.PRE_TOKEN_GENERATION,
+  userMigration: aws_cognito.UserPoolOperation.USER_MIGRATION,
+  verifyAuthChallengeResponse:
+    aws_cognito.UserPoolOperation.VERIFY_AUTH_CHALLENGE_RESPONSE,
+};
```

Comment:
Is there an escape hatch when a new trigger name is added? (idk if this happens).

Could we accept the actual value behind the CDK enum instead of remapping a string to the enum? That way a user could override and pass in any name if a new one is added.

Reply:
Ok, made it possible al ways around.

```ts
test("add a trigger to a user pool using on(aws_cognito.UserPoolOperation)", () => {
  const stack = new Stack();

  const pool = new UserPool(stack, "UserPool");

  pool.on(
    aws_cognito.UserPoolOperation.DEFINE_AUTH_CHALLENGE,
    new Function(
      stack,
      "CreateAuthChallenge",
      // must provide explicit event type
      async (event: CreateAuthChallengeTriggerEvent) => {
        return event;
      }
    )
  );
});

```

Reply:
This has the drawback of not inferring the type of the event. The enum is a synthetic enum where the enum items are objects (they are all the same type and can't be used as keys). So the user must annotate the `event: CreateAuthChallengeTriggerEvent ` type.